File: 백준 문제 풀이 모음/class 4/12851.py
import sys
input = sys.stdin.readline
n, m = map(int, input().split())

from collections import deque
graph = [-1] * 100001

def bfs (n) :
    queue = deque()
    queue.append([n])
    count = 0
    while queue :
        node = queue.popleft()
        for i in node :
            graph[i] = 1
        #print('node: ', node)

        if m in node :
            print(count)
            return node

        tmp = []
        for i in node :
            if 0 <= i-1 <= 100000 :
                if graph[i-1] != 1 :
                    tmp.append(i-1)
            if 0 <= i+1 <= 100000 :    
                if graph[i+1] != 1 :
                    tmp.append(i+1)
            if 0 <= 2*i <= 100000 :          
                if graph[2*i] != 1 :
                    tmp.append(2*i)

        count += 1        
        queue.append(tmp)

        
aaa = bfs(n)
total = 0
for i in range(len(aaa)) :
    if aaa[i] == m :
        total += 1
print(total)

File: 백준 문제 풀이 모음/kotlin/class 4/12851.kt


var visited = Array<Boolean>(0) { false }
var count = 0
var num = 0
fun main () = with(System.`in`.bufferedReader()) {
    val str = readLine().split(' ').toList()
    val n = str[0].toInt()
    val k = str[1].toInt()
    visited = Array<Boolean> (200001) { false }

    if(n==k) {
        println(0)
        println(1)
    } else {
        visited[n] = true
        val x = BFS(n, k)
        x.forEach {
            if (it == k ) {
                num++
            }
        }

//        println(x)
        println(count)
        println(num)

    }
}

fun BFS(n: Int, k : Int) : MutableList<Int> {
    var final_node = mutableListOf<Int>()
    val deque = ArrayDeque<MutableList<Int>>()
    deque.add(mutableListOf(n))


    while (deque[0].isNotEmpty()) {
//        println(deque)
        val node = deque.removeFirst()
        node.forEach {
            visited[it] = true
        }


        if (node.contains(k)) {
            final_node= node
            break
        }

        val tmp = mutableListOf<Int>()
        node.forEach {node ->
            val list = listOf<Int>(node * 2, node -1, node + 1)

            list.forEach {
                if (0<= it && it < 200001 && !visited[it]) {
                    tmp.add(it)
                }
            }
        }

        count ++
        deque.add(tmp)

    }

    return final_node
}